[test] fundrawtransaction: lock watch-only shared address

kallewoof opened the pull request:
`self.nodes[0]` creates an address which is watch-only-shared with `self.nodes[3]`. If `nodes[0]` spends the associated UTXO during any of its sends later, the watchonly test will fail, as `nodes[3]` now has insufficient funds.

I ran into this in #12257 and this commit is in that PR as well, but I figured I'd split it out (and remove from there once/if merged).

promag reviewed (commented):
Concept ACK.

Another approach is to use `listunspent --addresses [watchonly_address]`.

Either way maybe you could extract this "lock unspents by txid and address" to a separate function, it sure makes the test easier to read and might be useful elsewhere.

kallewoof committed:
[test] fundrawtransaction: lock watch-only shared address

self.nodes[0] creates an address which is watch-only-shared with self.nodes[3]. If nodes[0] spends the associated UTXO during any of its sends later, the watchonly test will fail, as nodes[3] now has insufficient funds.

Note that this also adds a new find_vout_for_address function to the test framework.

kallewoof commented:
@promag That's a good point. I moved parts of it into util as `find_vout_for_address`.

promag commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -472,6 +488,8 @@ def run_test(self):
         connect_nodes_bi(self.nodes,0,2)
         connect_nodes_bi(self.nodes,0,3)
+        # Again lock the watchonly UTXO or nodes[0] may spend it
+        self.nodes[0].lockunspent(False, [{"txid": watchonly_txid, "vout": watchonly_vout}])
```
Nit, add comment like "because restarting a node looses the locked unspents" (you may find better wording).

kallewoof replied:
Done!

promag commented on test/functional/test_framework/util.py:
```diff
@@ -569,3 +569,16 @@ def mine_large_block(node, utxos=None):
+    """
+    tx = node.getrawtransaction(txid, True)
+    for i in range(len(tx["vout"])):
+        o = tx["vout"][i]
```
Nit, inline `o` below (only used once).

kallewoof replied:
Good point. Fixed.

promag reviewed (commented):
ACK.

Just want to point that if `*rawtransaction` were used instead of `sendtoaddress` then the vout would be known.

laanwj commented:
utACK - can you please change the commit message of f3ebd07 to the format
```
<title (one line)>
<empty line>
<description... (possibly multiple lines)>
```
Otherwise tooling that parses the commit messages will make it into one long title.

kallewoof commented:
@laanwj Oops, didn't realize you needed an empty line. Fixed.

MarcoFalke commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -57,6 +67,12 @@ def run_test(self):
+
+        # Lock UTXO so nodes[0] doesn't accidentally spend it
+        watchonly_vout = find_vout_for_address(self.nodes[0], watchonly_txid, watchonly_address)
+        assert watchonly_vout != -1 # it must exist
```
You can move the assert into `find_vout_for_address`.

If you don't want to do that, note that python has no types, so you don't have to return an integer. `None` works just fine.

kallewoof replied:
You mean `assert watchonly_vout is not None` and not return anything in `find_v..`?

kallewoof replied:
Actually nevermind, I'll move the assert instead.

jnewbery commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -5,7 +5,17 @@
 
 from test_framework.test_framework import BitcoinTestFramework
-from test_framework.util import *
+from decimal import Decimal
```
nit: please place standard library imports above local project imports

jnewbery reviewed (commented):
Tested ACK 73099805915ad11b264bc22040015eb3ac137b2b with one nit.

Commits can be squashed IMO.

kallewoof commented:
@jnewbery Squashed. I kept them split to make it obvious to reviewers that I am adding a new utility function, not just fixing something in a specific test.

promag commented:
utACK 891beb0.

jnewbery commented on test/functional/test_framework/util.py:
```diff
@@ -569,3 +569,14 @@ def mine_large_block(node, utxos=None):
+    given address. Raises runtime error exception if not found.
+    """
+    tx = node.getrawtransaction(txid, True)
+    for i in range(len(tx["vout"])):
```
Alternatively:

```python
    for i, vout in enumerate(tx["vout"]):
        if any([addr == a for a in vout["scriptPubKey"]["addresses"]]):
            return i
```

kallewoof replied:
Thanks! It doesn't feel strictly necessary but it's marginally better. Will do if this needs a rebase or gets more requests for other changes.

jnewbery replied:
Agreed, no need to change the branch for just this change

jnewbery reviewed (commented):
Tested ACK 891beb0f8a09810b179e39a680b579c2f6516db7

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/12265/commits/891beb0f8a09810b179e39a680b579c2f6516db7
